website/src/blog/2021-09-27-hot-chocolate-12/2021-09-27-hot-chocolate-12.md

```diff
@@ -0,0 +1,768 @@
+
+In order to take advantage of the new cache pass down the DataLoader options so that we can inject them from the DI.
+
+```csharp
+public class CustomBatchDataLoader : BatchDataLoader<string, string>
+{
+    public CustomBatchDataLoader(IBatchScheduler batchScheduler, DataLoaderOptions options)
+        : base(batchScheduler, options)
+    {
+    }
+
+    protected override Task<IReadOnlyDictionary<string, string>> LoadBatchAsync(
+        IReadOnlyList<string> keys,
+        CancellationToken cancellationToken)
+}
+```
+
+With DataLoader now always pass down the options and the batch scheduled and we will inject you the new unified cache. If you do not pass down the options object then we will just use a cache per DataLoader.
+
+The DataLoader caches with Hot Chocolate 12 are also now pooled meaning that the cache object is cleaned, preserved and reused which will safe you memory. Further we have now introduced `DataLoaderDiagnosticEventListener` which allows you to monitor DataLoader execution.
+
+All diagnostic listener cannot be registered with the schema.
+
+```csharp
+services.AddGraphQLServer()
+    .AddDiagnosticEventListener<MyDataLoaderEventListener>()
+    ...
+```
+
+# Stream and Defer
+
+With Hot Chocolate 11 we introduced the `@defer` directive which allows you to defer parts of your query so that you get the most important data first and the execution of more expensive parts of your query is deprioritized.
+With Hot Chocolate 12 we now are introducing the `@stream` directive which allows you to take advantage of async enumerators and define how much data of a stream you want to get immediately and what shall be deferred to a later point in time.
+
+```graphql
+{
+  persons @stream(initialCount: 2) {
+    name
+  }
+}
+```
+
+Streamworks with any list in Hot Chocolate but it is only really efficient and gives you all the benefits of using stream if your resolver returns a `IAsyncEnumerable<T>`. Our internal paging middleware at the moment does not work with `IAsyncEnumerable<T>` which means you can stream the results, but you still will have the fill execution impact on the initial piece of the query. We however will rework pagination to use `IAsyncEnumerable<T>`when slicing the data and executing the query.
+
+Stream and defer both work with Banana Cake Pop if your browser is chrome based. We already have an update in the works to make it work with Safari. We will issue the BCP update with version 12.1.
+
+# ASP.NET Core improvements
+
+# The little things that will make your live easier
```

Comment:
The following items below will help improve developer experience using HotChocolate.